### DISCUSSION TITLE
Implement H2-nonconforming elements

### DISCUSSION THREAD

**Discussion by xida2020** [kinnala/scikit-fem#749] [General]
I find that ElementGlobal is a powerful class. I have ever programmed Argyris element in MATLAB, it is a tedious project. By contrast, ElementTriArgyris class in scikit-fem is concise and readable, it is superb job.  For fourth order problem, there are Morley element, Hermite element, and Argyris element over trianglular elements using ElementGlobal.  So I take a try for more elements, from the following book,

"Z. Shi, M. Wang, Finite Element Methods, Scientific Publishers, Beijing, 2013."

I find these H2-nonconforming elements for four order problem. they are Zienkiewicz element, Morley-Zienkiewicz element, Modified Zienkiewicz element, 12-parameter Triangle Plate Element and 15-parameter Triangle Plate Element, but only last one works.

I'm wondering why the others cannot be implemented by using ElementGlobal. I guess the reason is that the number of DOFs of Morley elment, Hermite element, 15-parameter Triangle Plate element and Argyris element is 6,10,15,21, corresponding to {1,x,y,x^2,xy,y^2},{1,x,y,x^2,xy,y^2,x^3,x^2y,xy^2,y^3},{1,x,y,x^2,xy,y^2,x^3,x^2y,xy^2,y^3,x^4,x^3y,x^2y^2,xy^3,y^4} and {1,x,y,x^2,xy,y^2,x^3,x^2y,xy^2,y^3,x^4,x^3y,x^2y^2,xy^3,y^4,x^5,x^4y,x^3y^2,x^2y^3,xy^4,y^5}.

My question is how to implement the rest elements.

**Comment by xida2020** [comment]
I could not write testcode and copy the codes for 15-parameter Triangle Plate Element as follow.  and I want to know whether it is correct. 

class ElementTri15ParameterPlate(ElementGlobal):
    """15-parameter Triangle Plate Element"""

    nodal_dofs = 3
    facet_dofs = 2
    maxdeg = 4
    dofnames = ['u', 'u_x', 'u_y','u', 'u_n']
    doflocs = np.array([[0., 0.],
                        [0., 0.],
                        [0., 0.],
                        [1., 0.],
                        [1., 0.],
                        [1., 0.],
                        [0., 1.],
                        [0., 1.],
                        [0., 1.],
                        [.5, 0.],
                        [.5, .5],
                        [0., .5],
                        [.5, 0.],
                        [.5, .5],
                        [0., .5]])

    refdom = RefTri

    def gdof(self, F, w, i):
        if i == 0:
            return F[()](*w['v'][0])
        elif i == 1:
            return F[(0,)](*w['v'][0])
        elif i == 2:
            return F[(1,)](*w['v'][0]) 
        elif i == 3:
            return F[()](*w['v'][1]) 
        elif i == 4:
            return F[(0,)](*w['v'][1]) 
        elif i == 5:
            return F[(1,)](*w['v'][1]) 
        elif i == 6:
            return F[()](*w['v'][2]) 
        elif i == 7:
            return F[(0,)](*w['v'][2])
        elif i == 8:
            return F[(1,)](*w['v'][2])
        elif i == 9:
            return F[()](*w['e'][0]) 
        elif i == 10:
            return F[()](*w['e'][1])
        elif i == 11:
            return F[()](*w['e'][2])
        elif i == 12:
            return (F[(0,)](*w['e'][0]) * w['n'][0, 0]
                    + F[(1,)](*w['e'][0]) * w['n'][0, 1])
        elif i == 13:
            return (F[(0,)](*w['e'][1]) * w['n'][1, 0]
                    + F[(1,)](*w['e'][1]) * w['n'][1, 1])
        elif i == 14:
            return (F[(0,)](*w['e'][2]) * w['n'][2, 0]
                    + F[(1,)](*w['e'][2]) * w['n'][2, 1])
        self._index_error()

**Comment by xida2020** [comment]
class ElementTriMZ(ElementGlobal):
    """The modified Zienkiewicz element."""

    nodal_dofs = 3
    facet_dofs = 0
    maxdeg = 3
    tensorial_basis = True
    dofnames = ['u', 'u_x', 'u_y']
    doflocs = np.array([[0., 0.],
                        [0., 0.],
                        [0., 0.],
                        [1., 0.],
                        [1., 0.],
                        [1., 0.],
                        [0., 1.],
                        [0., 1.],
                        [0., 1.]])
    refdom = RefTri

    def gdof(self, F, w, i):
        if i == 0:
            return F[()](*w['v'][0])
        elif i == 1:
            return F[(0,)](*w['v'][0])
        elif i == 2:
            return F[(1,)](*w['v'][0]) 
        elif i == 3:
            return F[()](*w['v'][1]) 
        elif i == 4:
            return F[(0,)](*w['v'][1]) 
        elif i == 5:
            return F[(1,)](*w['v'][1]) 
        elif i == 6:
            return F[()](*w['v'][2]) 
        elif i == 7:
            return F[(0,)](*w['v'][2])
        elif i == 8:
            return F[(1,)](*w['v'][2])
        self._index_error()

The codes above is for modified Zienkiewicz element, but it doesn't work.

**Comment by xida2020** [comment]
class ElementTriMorleyZienkiewicz(ElementGlobal):
    """The Morley-Zienkiewicz element."""

    nodal_dofs = 3
    facet_dofs = 1
    maxdeg = 4
    dofnames = ['u', 'u_x', 'u_y', 'u_n']
    doflocs = np.array([[0., 0.],
                        [0., 0.],
                        [0., 0.],
                        [1., 0.],
                        [1., 0.],
                        [1., 0.],
                        [0., 1.],
                        [0., 1.],
                        [0., 1.],
                        [.5, 0.],
                        [.5, .5],
                        [0., .5]])

    refdom = RefTri

    def gdof(self, F, w, i):
        if i == 0:
            return F[()](*w['v'][0])
        elif i == 1:
            return F[(0,)](*w['v'][0])
        elif i == 2:
            return F[(1,)](*w['v'][0]) 
        elif i == 3:
            return F[()](*w['v'][1]) 
        elif i == 4:
            return F[(0,)](*w['v'][1]) 
        elif i == 5:
            return F[(1,)](*w['v'][1]) 
        elif i == 6:
            return F[()](*w['v'][2]) 
        elif i == 7:
            return F[(0,)](*w['v'][2])
        elif i == 8:
            return F[(1,)](*w['v'][2])
        elif i == 9:
            return (F[(0,)](*w['e'][0]) * w['n'][0, 0]
                    + F[(1,)](*w['e'][0]) * w['n'][0, 1])
        elif i == 10:
            return (F[(0,)](*w['e'][1]) * w['n'][1, 0]
                    + F[(1,)](*w['e'][1]) * w['n'][1, 1])
        elif i == 11:
            return (F[(0,)](*w['e'][2]) * w['n'][2, 0]
                    + F[(1,)](*w['e'][2]) * w['n'][2, 1])
        self._index_error()

The codes above is for Morley-Zienkiewicz element, but it also doesn't work.

**Comment by kinnala** [comment]
I'll take a look if I can figure out what is the issue.

**Comment by kinnala** [comment]
I changed the DOF order in 15 parameter plate element and now it seems to work better:
```python
from skfem import *
from skfem.element.element_global import ElementGlobal
from skfem.refdom import RefTri
from skfem.helpers import *
import numpy as np

class ElementTri15ParameterPlate(ElementGlobal):
    """15-parameter Triangle Plate Element"""

    nodal_dofs = 3
    facet_dofs = 2
    maxdeg = 4
    dofnames = ['u', 'u_x', 'u_y', 'u', 'u_n']
    doflocs = np.array([[0., 0.],
                        [0., 0.],
                        [0., 0.],
                        [1., 0.],
                        [1., 0.],
                        [1., 0.],
                        [0., 1.],
                        [0., 1.],
                        [0., 1.],
                        [.5, 0.],
                        [.5, .5],
                        [0., .5],
                        [.5, 0.],
                        [.5, .5],
                        [0., .5]])

    refdom = RefTri

    def gdof(self, F, w, i):
        if i == 0:
            return F[()](*w['v'][0])
        elif i == 1:
            return F[(0,)](*w['v'][0])
        elif i == 2:
            return F[(1,)](*w['v'][0]) 
        elif i == 3:
            return F[()](*w['v'][1]) 
        elif i == 4:
            return F[(0,)](*w['v'][1]) 
        elif i == 5:
            return F[(1,)](*w['v'][1]) 
        elif i == 6:
            return F[()](*w['v'][2]) 
        elif i == 7:
            return F[(0,)](*w['v'][2])
        elif i == 8:
            return F[(1,)](*w['v'][2])
        elif i == 9:
            return F[()](*w['e'][0]) 
        elif i == 10:
            return (F[(0,)](*w['e'][0]) * w['n'][0, 0]
                    + F[(1,)](*w['e'][0]) * w['n'][0, 1])
        elif i == 11:
            return F[()](*w['e'][1])
        elif i == 12:
            return (F[(0,)](*w['e'][1]) * w['n'][1, 0]
                    + F[(1,)](*w['e'][1]) * w['n'][1, 1])
        elif i == 13:
            return F[()](*w['e'][2])
        elif i == 14:
            return (F[(0,)](*w['e'][2]) * w['n'][2, 0]
                    + F[(1,)](*w['e'][2]) * w['n'][2, 1])
        self._index_error()



m = MeshTri().refined(3)


ib = Basis(m, ElementTri15ParameterPlate())

t = 1.
E = 1.
nu = 0.3
D = E * t ** 3 / (12. * (1. - nu ** 2))

@BilinearForm
def bilinf(u, v, w):

    def C(T):
        trT = T[0, 0] + T[1, 1]
        return E / (1. + nu) * \
            np.array([[T[0, 0] + nu / (1. - nu) * trT, T[0, 1]],
                      [T[1, 0], T[1, 1] + nu / (1. - nu) * trT]])

    return t ** 3 / 12.0 * ddot(C(dd(u)), dd(v))

def load(x):
    return np.sin(np.pi * x[0]) * np.sin(np.pi * x[1])

@LinearForm
def linf(v, w):
    return load(w.x) * v

K = asm(bilinf, ib)
f = asm(linf, ib)

x = solve(*condense(K, f, D=ib.get_dofs().all('u')))

from skfem.visuals.matplotlib import *

plot(ib, x, Nrefs=3)
show()
```
![Figure_1](https://user-images.githubusercontent.com/973268/133931291-9fe8b2c3-8e40-4660-8bbc-fe9d4d2d289c.png)

**Comment by kinnala** [comment]
I think the other methods use a different type of polynomial basis. E.g., Bell triangle does not have full polynomial space and I have not looked into how ElementGlobal should be modified.

**Comment by kinnala** [comment]
Should we add 15 parameter nonconforming triangle to scikit-fem?

**Comment by xida2020** [comment]
Sure, we have more choices when solving four order problems

**Comment by xida2020** [comment]
For the modified Zienkiewicz element, a H2-nonconforming elment, can't be worked by using ElementGlobal class. And now I do another try. I write it in explicit basis and modify ElementH1 class to get a new  ElementH2 class, with adding  a hessian operator.

import numpy as np
from ..element_h2 import ElementH2 

from ...refdom import RefTri


class ElementTriMZ(ElementH2):
    """The modified Zienkiewicz element."""

    nodal_dofs = 3
    facet_dofs = 0
    maxdeg = 3
    dofnames = ['u', 'u_x', 'u_y']
    doflocs = np.array([[0., 0.],
                        [0., 0.],
                        [0., 0.],
                        [1., 0.],
                        [1., 0.],
                        [1., 0.],
                        [0., 1.],
                        [0., 1.],
                        [0., 1.]])
    refdom = RefTri

    def lbasis(self, X, i):
        x, y = X

        if i == 0:                
            phi = -(x + y - 1)*(- 12*x**2*y - 2*x**2 - 12*x*y**2 + 8*x*y + x - 2*y**2 + y + 1)
            dphi = np.array([
                36*x**2*y + 6*x**2 + 48*x*y**2 - 36*x*y - 6*x + 12*y**3 - 18*y**2 + 6*y,
                12*x**3 + 48*x**2*y - 18*x**2 + 36*x*y**2 - 36*x*y + 6*x + 6*y**2 - 6*y
            ])
            d2phi = np.array([
                12*x - 36*y + 72*x*y + 48*y**2 - 6,
                36*x**2 + 96*x*y - 36*x + 36*y**2 - 36*y + 6,
                12*y - 36*x + 72*x*y + 48*x**2 - 6
            ])
        elif i == 1:
            phi = (x*(x + y - 1)*(2*x + y - 2))/2 + x*y*(x + y - 1)*(2*x + y - 1) + (x*y*(12*y - 6)*(x + y - 1))/4
            dphi = np.array([
                6*x**2*y + 3*x**2 + 12*x*y**2 - 6*x*y - 4*x + 4*y**3 - 6*y**2 + y + 1,
                x*(2*x**2 + 12*x*y - 3*x + 12*y**2 - 12*y + 1)
            ])
            d2phi = np.array([
                6*x - 6*y + 12*x*y + 12*y**2 - 4,
                6*x**2 + 24*x*y - 6*x + 12*y**2 - 12*y + 1,
                x*(12*x + 24*y - 12)
            ])
        elif i == 2:
            phi = (y*(x + y - 1)*(x + 2*y - 2))/2 + x*y*(x + y - 1)*(x + 2*y - 1) + (x*y*(12*x - 6)*(x + y - 1))/4
            dphi = np.array([
                y*(12*x**2 + 12*x*y - 12*x + 2*y**2 - 3*y + 1),
                4*x**3 + 12*x**2*y - 6*x**2 + 6*x*y**2 - 6*x*y + x + 3*y**2 - 4*y + 1
            ])
            d2phi = np.array([
                y*(24*x + 12*y - 12),
                12*x**2 + 24*x*y - 12*x + 6*y**2 - 6*y + 1,
                6*y - 6*x + 12*x*y + 12*x**2 - 4
            ])
        elif i == 3:
            phi = -x*(6*x**2*y + 2*x**2 + 12*x*y**2 - 9*x*y - 3*x + 6*y**3 - 9*y**2 + 3*y)
            dphi = np.array([
                - 18*x**2*y - 6*x**2 - 24*x*y**2 + 18*x*y + 6*x - 6*y**3 + 9*y**2 - 3*y,
                -x*(6*x**2 + 24*x*y - 9*x + 18*y**2 - 18*y + 3)
            ])
            d2phi = np.array([
                18*y - 12*x - 36*x*y - 24*y**2 + 6,
                - 18*x**2 - 48*x*y + 18*x - 18*y**2 + 18*y - 3,
                -x*(24*x + 36*y - 18)
            ])
        elif i == 4:
            phi = (x*(6*x**2*y + 2*x**2 + 12*x*y**2 - 9*x*y - 2*x + 6*y**3 - 9*y**2 + 3*y))/2
            dphi = np.array([
                9*x**2*y + 3*x**2 + 12*x*y**2 - 9*x*y - 2*x + 3*y**3 - (9*y**2)/2 + (3*y)/2,
                (x*(6*x**2 + 24*x*y - 9*x + 18*y**2 - 18*y + 3))/2
            ])
            d2phi = np.array([
                6*x - 9*y + 18*x*y + 12*y**2 - 2,
                9*x**2 + 24*x*y - 9*x + 9*y**2 - 9*y + 3/2,
                (x*(24*x + 36*y - 18))/2
            ])
        elif i == 5:
            phi = (x*y*(- 2*x**2 + 3*x + 2*y**2 - 3*y + 1))/2
            dphi = np.array([
                (y*(- 6*x**2 + 6*x + 2*y**2 - 3*y + 1))/2,
                (x*(- 2*x**2 + 3*x + 6*y**2 - 6*y + 1))/2
            ])
            d2phi = np.array([
                -3*y*(2*x - 1),
                - 3*x**2 + 3*x + 3*y**2 - 3*y + 1/2,
                3*x*(2*y - 1)
            ])
        elif i == 6:
            phi = -y*(6*x**3 + 12*x**2*y - 9*x**2 + 6*x*y**2 - 9*x*y + 3*x + 2*y**2 - 3*y)
            dphi = np.array([
                -y*(18*x**2 + 24*x*y - 18*x + 6*y**2 - 9*y + 3),
                - 6*x**3 - 24*x**2*y + 9*x**2 - 18*x*y**2 + 18*x*y - 3*x - 6*y**2 + 6*y
            ])
            d2phi = np.array([
                -y*(36*x + 24*y - 18),
                - 18*x**2 - 48*x*y + 18*x - 18*y**2 + 18*y - 3,
                18*x - 12*y - 36*x*y - 24*x**2 + 6
            ])
        elif i == 7:
            phi = (x*y*(2*x**2 - 3*x - 2*y**2 + 3*y + 1))/2
            dphi = np.array([
                (y*(6*x**2 - 6*x - 2*y**2 + 3*y + 1))/2,
                (x*(2*x**2 - 3*x - 6*y**2 + 6*y + 1))/2
            ])
            d2phi = np.array([
                3*y*(2*x - 1),
                3*x**2 - 3*x - 3*y**2 + 3*y + 1/2,
                -3*x*(2*y - 1)
            ])
        elif i == 8:
            phi = (y*(6*x**3 + 12*x**2*y - 9*x**2 + 6*x*y**2 - 9*x*y + 3*x + 2*y**2 - 2*y))/2
            dphi = np.array([
                (y*(18*x**2 + 24*x*y - 18*x + 6*y**2 - 9*y + 3))/2,
                3*x**3 + 12*x**2*y - (9*x**2)/2 + 9*x*y**2 - 9*x*y + (3*x)/2 + 3*y**2 - 2*y
            ])
            d2phi = np.array([
                (y*(36*x + 24*y - 18))/2,
                9*x**2 + 24*x*y - 9*x + 9*y**2 - 9*y + 3/2,
                6*y - 9*x + 18*x*y + 12*x**2 - 2
            ])
        else:
            self._index_error()

        return phi, dphi, d2phi

**Comment by xida2020** [comment]
I am not sure whether this code is correct. And I compared Hermite element using the same way with ElementTriHermite in scikit-fem. I copy the code of ElementTriHermite with explicit basis here:

import numpy as np

from ..element_h2 import ElementH2
from ...refdom import RefTri


class ElementTriHermite0(ElementH2):
    """The Hermite element with 3 DOFs per vertex and one interior DOF."""

    nodal_dofs = 3
    interior_dofs = 1
    maxdeg = 3
    dofnames = ['u', 'u_x', 'u_y', 'u']
    doflocs = np.array([[0., 0.],
                        [0., 0.],
                        [0., 0.],
                        [1., 0.],
                        [1., 0.],
                        [1., 0.],
                        [0., 1.],
                        [0., 1.],
                        [0., 1.],
                        [1 / 3, 1 / 3]])
    refdom = RefTri

    def lbasis(self, X, i):
        x, y = X

        if i == 0:                
            phi =  (1-x-y)**2*(3-2*(1-x-y))-7*x *y *(1-x-y)
            dphi = np.array([
                6*x**2 + 26*x *y - 6*x + 13*y**2 - 13*y,
                13*x**2 + 26*x *y - 13*x + 6*y**2 - 6*y
            ])
            d2phi = np.array([
                12*x + 26*y - 6,
                26*x + 26*y - 13,
                26*x + 12*y - 6
            ])
        elif i == 1:
            phi = x *(1-x-y) *((1-x-y)-y)
            dphi = np.array([
                3*x**2 + 6*x *y - 4*x + 2*y**2 - 3*y + 1,
                x *(3*x + 4*y - 3)
            ])
            d2phi = np.array([
                6*x + 6*y - 4,
                6*x + 4*y - 3,
                4*x
            ])
        elif i == 2:
            phi = y *(1-x-y) *((1-x-y)-x)
            dphi = np.array([
                y *(4*x + 3*y - 3),
                2*x**2 + 6*x *y - 3*x + 3*y**2 - 4*y + 1
            ])
            d2phi = np.array([
                4*y,
                4*x + 6*y - 3,
                6*x + 6*y - 4
            ])
        elif i == 3:
            phi = x**2*(3-2*x)-7*x *y *(1-x-y) 
            dphi = np.array([
                - 6*x**2 + 14*x *y + 6*x + 7*y**2 - 7*y,
                7*x *(x + 2*y - 1)
            ])
            d2phi = np.array([
                14*y - 12*x + 6,
                14*x + 14*y - 7,
                14*x
            ])
        elif i == 4:
            phi = x *(2*y *(1-x-y)+x**2-x)
            dphi = np.array([
                3*x**2 - 4*x *y - 2*x - 2*y**2 + 2*y,
                -x *(2*x + 4*y - 2)
            ])
            d2phi = np.array([
                6*x - 4*y - 2,
                2 - 4*y - 4*x,
                -4*x
            ])
        elif i == 5:
            phi = x *y *(x-(1.-x-y))
            dphi = np.array([
                y *(4*x + y - 1),
                x *(2*x + 2*y - 1)
            ])
            d2phi = np.array([
                4*y,
                4*x + 2*y - 1,
                2*x
            ])
        elif i == 6:
            phi = y**2*(3-2*y)-7*x *y *(1-x-y)
            dphi = np.array([
                7*y *(2*x + y - 1),
                7*x**2 + 14*x *y - 7*x - 6*y**2 + 6*y
            ])
            d2phi = np.array([
                14*y,
                14*x + 14*y - 7,
                14*x - 12*y + 6
            ])
        elif i == 7:
            phi = y *(2*x *(1-x-y)+y**2-y)
            dphi = np.array([
                -y *(4*x + 2*y - 2),
                - 2*x**2 - 4*x *y + 2*x + 3*y**2 - 2*y
            ])
            d2phi = np.array([
                -4*y,
                2 - 4*y - 4*x,
                6*y - 4*x - 2
            ])
        elif i == 8:
            phi = x *y *(y-(1. -x-y))
            dphi = np.array([
                y *(2*x + 2*y - 1),
                x *(x + 4*y - 1)
            ])
            d2phi = np.array([
                2*y,
                2*x + 4*y - 1,
                4*x
            ])
        elif i == 9:
            phi = 27*x *y *(1-x-y)
            dphi = np.array([
                -27*y *(2*x + y - 1),
                -27*x *(x + 2*y - 1)
            ])
            d2phi = np.array([
                -54*y,
                27 - 54*y - 54*x,
                -54*x
            ])
        else:
            self._index_error()

        return phi, dphi, d2phi

**Comment by xida2020** [comment]
And they are different to solve the same problem, could you help me look at  it and check it? Thank you in advance.
